lib/logflare/alerting.ex

```diff
@@ -0,0 +1,242 @@
+
+  @doc """
+  Updates or creates a new Citrine.Job based on a given AlertQuery
+  """
+  @spec upsert_alert_job(AlertQuery.t()) :: {:ok, Citrine.Job.t()}
+  def upsert_alert_job(%AlertQuery{} = alert_query) do
+    Logflare.AlertsScheduler.put_job(%Citrine.Job{
+      id: alert_query.id,
+      schedule: alert_query.cron,
+      extended_syntax: false,
+      task: {:run_alert, [alert_query]}
+    })
+
+    {:ok, get_alert_job(alert_query)}
+  end
+
+  @doc """
+  Initializes and ensures that all alert jobs are created.
+  TODO: batching instead of loading whole table.
+  """
+  def init_alert_jobs do
+    AlertQuery
+    |> Repo.all()
+    |> Stream.each(fn alert_query ->
+      if get_alert_job(alert_query) == nil do
+        upsert_alert_job(alert_query)
+      end
+    end)
+    |> Stream.run()
+
+    :ok
+  end
+
+  @doc """
+  Performs the check lifecycle of an AlertQuery.
+
+  Send notifications if necessary configurations are set. If no results are returned from the query execution, no alert is sent.
+  """
+  @spec run_alert(AlertQuery.t()) :: :ok
+  def run_alert(%AlertQuery{} = alert_query) do
+    alert_query = alert_query |> Repo.preload([:user])
+
+    with {:ok, [_ | _] = results} <- execute_alert_query(alert_query) do
+      if alert_query.webhook_notification_url do
+        WebhookAdaptor.Client.send(alert_query.webhook_notification_url, %{
+          "result" => results
+        })
+      end
+
+      if alert_query.slack_hook_url do
+        SlackAdaptor.send_message(alert_query.slack_hook_url, results)
+      end
+
+      :ok
+    else
+      {:ok, []} ->
+        :ok
+
+      other ->
+        other
+    end
+  end
+
+  @doc """
+  Deletes an AlertQuery's Citrine.Job from the scheduler
+  noop if already deleted.
+
+  ### Examples
+    iex> delete_alert_job(%AlertQuery{})
+    :ok
+    iex> delete_alert_job(alert_query.id)
+    :ok
+  """
+  @spec delete_alert_job(AlertQuery.t() | number()) :: :ok
+  def delete_alert_job(%AlertQuery{id: id}), do: delete_alert_job(id)
+
+  def delete_alert_job(alert_id) do
+    Logflare.AlertsScheduler.delete_job(alert_id)
+  end
+
+  @doc """
+  Executes an AlertQuery and returns its results
+
+  Requires `:user` key to be preloaded.
+
+  ### Examples
+    iex> execute_alert_query(alert_query)
+    {:ok, [{"user_id" => "my-user-id"}]}
+  """
+  @spec execute_alert_query(AlertQuery.t()) :: {:ok, [map()]}
+  def execute_alert_query(%AlertQuery{user: %User{}} = alert_query) do
+    Logger.info("Executing AlertQuery | #{alert_query.name} | #{alert_query.id}")
+
+    with {:ok, transformed_query} <-
+           Logflare.Sql.transform(:bq_sql, alert_query.query, alert_query.user_id),
+         {:ok, %{rows: rows}} <-
+           Logflare.BqRepo.query_with_sql_and_params(
+             alert_query.user,
+             alert_query.user.bigquery_project_id || env_project_id(),
+             transformed_query,
+             [],
+             parameterMode: "NAMED",
+             maxResults: 1000,
```

Comment:
Only thing is maybe limit maxResults to 5 because these will be displayed somewhere.  We currently limit the Slack alerts to the latest 5 log events because they're just too much usually.

Reply:
Restricting it at the query level makes it much less flexible, as being able to send such events to a webhook can be quite and interesting use case.
Also, it's up to the user to fine tune their query to make the display optimal, and I would expect most queries to be doing aggregates.

We could add in more docs and best practices around it instead

Reply:
Maybe just make it overridable like we do with Endpoints and default it to 5 on the form. 

I guarantee the first alert someone gets with 1000 things is going to be WTF because it happened with the current alerts.

If they want more they can change it there.